# 🧾 Table of Contents

- [2. Getting the Data](#2.-Getting-the-Data)
  - [Overviewing the Data](#Overviewing-the-Data)
  - [Preparing our toolbox](#Preparing-our-toolbox)
    - [Application_train.csv](#Application_train.csv)
      - [Highlights & Ideas](#Highlights-&-Ideas)
    - [bureau.csv](#bureau.csv)
      - [Highlights & Ideas](#Highlights-&-Ideas)
    - [bureau_balance.csv](#bureau_balance.csv)
      - [Highlights & Ideas](#Highlights-&-Ideas)
    - [previous_application.csv](#previous_application.csv)
      - [Highlights & Ideas](#Highlights-&-Ideas)
    - [POS_CASH_balance.csv](#POS_CASH_balance.csv)
      - [Highlights & Ideas](#Highlights-&-Ideas)
    - [installments_payments.csv](#installments_payments.csv)
      - [Highlights & Ideas](#Highlights-&-Ideas)
    - [credit_card_balance.csv](#credit_card_balance.csv)
      - [Highlights & Ideas](#Highlights-&-Ideas)
    - [Final Thoughts](#Final-Thoughts)

# 2. Getting the Data

The available data is in this [link](https://drive.google.com/file/d/17fyteuN2MdGdbP5_Xq_sySN_yH91vTup/view). It is a ```.zip``` file with 673mb. There are no impediments related to legal obligations, authorizations or sensitive information.

With a file of this size, we will definetely experience restrictions on uploading it to Github. So, I either use Git Large File Storage solution or automate the process of getting the data using ```make```. I chose the latter to improve reproducibility and leave the project structure as light as possible. To use it:

1. The Shell Terminal in the root folder of the project:
 
    ```make data```

 1. The Ipython Terminal Interpreter from this notebook:
 
    ```!cd .. && make data```

The package requirements will be installed and the data will be download into ```/data/raw/``` folder.

## Overviewing the Data

There are 10 available Datasets and a Metadata PDF file. From the 10 Datasets:
- 1 is the Columns Descriptions.
- 1 is a Sample Submission.
- 1 is the test set for the Final Submission.
- The rest is data to be explored.
  
- ER Diagram below:

<div align="center">
<img src="../references/home_credit.png" alt="Home Credit ER Diagram" style="height: 500px"/>
</div>

Ok, now it is time to take a high-level view into the datasets.

## Preparing our toolbox

In [9]:
%load_ext autoreload
%autoreload 2

from src.data.explore_data import (
    list_datasets,
    describe_feature,
    overview_data,
    create_dataframe,
    describe_features,
)

The autoreload extension is already loaded. To reload it, use:
  %reload_ext autoreload


In [5]:
list_datasets()

('application_test_student.csv',
 'HomeCredit_columns_description.csv',
 'POS_CASH_balance.csv',
 'credit_card_balance.csv',
 'installments_payments.csv',
 'application_train.csv',
 'bureau.csv',
 'previous_application.csv',
 'bureau_balance.csv',
 'sample_submission.csv')

### Application_train.csv
*Starting with the flagship dataset*

In [10]:
overview_data("application_train.csv", sample_mode="random", sample_size=5)

#### Dataset name (.csv)

application_train.csv


#### Dataset Random Samples

SK_ID_CURR  TARGET NAME_CONTRACT_TYPE CODE_GENDER FLAG_OWN_CAR  \
161368      412052       0         Cash loans           F            N   
177644      391895       0         Cash loans           M            Y   
42295       279443       0         Cash loans           F            N   
161674      237294       0         Cash loans           F            N   
235457      401672       0         Cash loans           M            N   

       FLAG_OWN_REALTY  CNT_CHILDREN  AMT_INCOME_TOTAL  AMT_CREDIT  \
161368               N             1          135000.0    755190.0   
177644               Y             1          121500.0    621900.0   
42295                Y             0          135000.0    651595.5   
161674               N             0           99000.0    334152.0   
235457               N             1           90000.0    319981.5   

        AMT_ANNUITY  AMT_GOODS_PRICE NAME_TYPE_SUITE NAME_INCOME_TYPE  \
161368      36459.0         675000.0   Unaccompanied          Working   
177644      49266.0         562500.0   Unaccompanied          Working   
42295       33399.0         526500.0   Unaccompanied          Working   
161674      14157.0         270000.0   Unaccompanied        Pensioner   
235457      12316.5         243000.0   Unaccompanied          Working   

                  NAME_EDUCATION_TYPE    NAME_FAMILY_STATUS  \
161368               Higher education               Married   
177644  Secondary / secondary special               Married   
42295                Higher education  Single / not married   
161674  Secondary / secondary special               Married   
235457  Secondary / secondary special               Married   

        NAME_HOUSING_TYPE  REGION_POPULATION_RELATIVE  DAYS_BIRTH  \
161368   Rented apartment                    0.018801      -10343   
177644  House / apartment                    0.030755      -11080   
42295   House / apartment                    0.024610      -10779   
161674  House / apartment                    0.009175      -21365   
235457  House / apartment                    0.005144      -13078   

        DAYS_EMPLOYED  DAYS_REGISTRATION  DAYS_ID_PUBLISH  OWN_CAR_AGE  \
161368           -370            -4547.0            -1583          NaN   
177644           -635            -3896.0            -3722         18.0   
42295            -482               -7.0            -3451          NaN   
161674         365243           -10459.0            -4599          NaN   
235457          -4128            -4192.0            -4190          NaN   

        FLAG_MOBIL  FLAG_EMP_PHONE  FLAG_WORK_PHONE  FLAG_CONT_MOBILE  \
161368           1               1                0                 1   
177644           1               1                0                 1   
42295            1               1                0                 1   
161674           1               0                0                 1   
235457           1               1                0                 1   

        FLAG_PHONE  FLAG_EMAIL     OCCUPATION_TYPE  CNT_FAM_MEMBERS  \
161368           0           0         Accountants              3.0   
177644           1           0             Drivers              3.0   
42295            0           0            Laborers              1.0   
161674           1           0                 NaN              2.0   
235457           0           0  Low-skill Laborers              3.0   

        REGION_RATING_CLIENT  REGION_RATING_CLIENT_W_CITY  \
161368                     2                            2   
177644                     2                            2   
42295                      2                            2   
161674                     2                            2   
235457                     2                            2   

       WEEKDAY_APPR_PROCESS_START  HOUR_APPR_PROCESS_START  \
161368                    TUESDAY                       11   
177644                    TUESDAY                       12   
42295                    THURSDAY            

#### Dataset info

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 246008 entries, 0 to 246007
Columns: 122 entries, SK_ID_CURR to AMT_REQ_CREDIT_BUREAU_YEAR
dtypes: float64(65), int64(41), object(16)
memory usage: 229.0+ MB


#### Dataset descriptive statistics

SK_ID_CURR         TARGET NAME_CONTRACT_TYPE CODE_GENDER  \
count   246008.000000  246008.000000             246008      246008   
unique            NaN            NaN                  2           3   
top               NaN            NaN         Cash loans           F   
freq              NaN            NaN             222622      161867   
mean    278280.072908       0.081176                NaN         NaN   
std     102790.909988       0.273106                NaN         NaN   
min     100002.000000       0.000000                NaN         NaN   
25%     189165.500000       0.000000                NaN         NaN   
50%     278392.500000       0.000000                NaN         NaN   
75%     367272.250000       0.000000                NaN         NaN   
max     456255.000000       1.000000                NaN         NaN   

       FLAG_OWN_CAR FLAG_OWN_REALTY   CNT_CHILDREN  AMT_INCOME_TOTAL  \
count        246008          246008  246008.000000      2.460080e+05   
unique            2               2            NaN               NaN   
top               N               Y            NaN               NaN   
freq         162355          170535            NaN               NaN   
mean            NaN             NaN       0.415527      1.689122e+05   
std             NaN             NaN       0.719922      2.603818e+05   
min             NaN             NaN       0.000000      2.565000e+04   
25%             NaN             NaN       0.000000      1.125000e+05   
50%             NaN             NaN       0.000000      1.485000e+05   
75%             NaN             NaN       1.000000      2.025000e+05   
max             NaN             NaN      19.000000      1.170000e+08   

          AMT_CREDIT    AMT_ANNUITY  AMT_GOODS_PRICE NAME_TYPE_SUITE  \
count   2.460080e+05  245998.000000     2.457820e+05          244960   
unique           NaN            NaN              NaN               7   
top              NaN            NaN              NaN   Unaccompanied   
freq             NaN            NaN              NaN          198726   
mean    5.996283e+05   27129.162648     5.389289e+05             NaN   
std     4.030672e+05   14504.965232     3.699738e+05             NaN   
min     4.500000e+04    1615.500000     4.050000e+04             NaN   
25%     2.700000e+05   16561.125000     2.385000e+05             NaN   
50%     5.147775e+05   24930.000000     4.500000e+05             NaN   
75%     8.086500e+05   34599.375000     6.795000e+05             NaN   
max     4.050000e+06  258025.500000     4.050000e+06             NaN   

       NAME_INCOME_TYPE            NAME_EDUCATION_TYPE NAME_FAMILY_STATUS  \
count            246008                         246008             246008   
unique                8                              5                  6   
top             Working  Secondary / secondary special            Married   
freq             126919                         174831             157074   
mean                NaN                            NaN                NaN   
std                 NaN                            NaN                NaN   
min                 NaN                            NaN                NaN   
25%                 NaN                            NaN                NaN   
50%                 NaN                            NaN                NaN   
75%                 NaN                            NaN                NaN   
max                 NaN                            NaN                NaN   

        NAME_HOUSING_TYPE  REGION_POPULATION_RELATIVE     DAYS_BIRTH  \
count              246008               246008.000000  246008.000000   
unique                  6                         NaN            NaN   
top     House / apartment                         NaN            NaN   
freq               218258                         NaN            NaN   
mean                  NaN                    0.020882  -16042.794393   
std                   NaN                    0.013852    4365.973763   

In [12]:
describe_features("application_train.csv", display_option="condensed")

Dataset                      Column  \
0    application_{train|test}.csv                  SK_ID_CURR   
1    application_{train|test}.csv                      TARGET   
2    application_{train|test}.csv          NAME_CONTRACT_TYPE   
3    application_{train|test}.csv                 CODE_GENDER   
4    application_{train|test}.csv                FLAG_OWN_CAR   
..                            ...                         ...   
117  application_{train|test}.csv   AMT_REQ_CREDIT_BUREAU_DAY   
118  application_{train|test}.csv  AMT_REQ_CREDIT_BUREAU_WEEK   
119  application_{train|test}.csv   AMT_REQ_CREDIT_BUREAU_MON   
120  application_{train|test}.csv   AMT_REQ_CREDIT_BUREAU_QRT   
121  application_{train|test}.csv  AMT_REQ_CREDIT_BUREAU_YEAR   

                                           Description Special  
0                             ID of loan in our sample     NaN  
1    Target variable (1 - client with payment diffi...     NaN  
2          Identification if loan is cash or revolving     NaN  
3                                 Gender of the client     NaN  
4                        Flag if the client owns a car     NaN  
..                                                 ...     ...  
117  Number of enquiries to Credit Bureau about the...     NaN  
118  Number of enquiries to Credit Bureau about the...     NaN  
119  Number of enquiries to Credit Bureau about the...     NaN  
120  Number of enquiries to Credit Bureau about the...     NaN  
121  Number of enquiries to Credit Bureau about the...     NaN  

[122 rows x 4 columns]

#### Highlights & Ideas

There are **246k+** records and **122** columns. A sizeable dataset.

- ```NAME_CONTRACT_TYPE```: research about types of loan contracts.
- From column 21 ```OWN_CAR_AGE``` to column 27 ```FLAG_EMAIL```, there is room for studying correlation and feature engineering, such as rearrangement.
- From column 44 ```APARTMENTS_AVG``` to column 90 ```EMERGENCYSTATE_MODE``` we get normalized aspects about the building where the client lives.
    - Their description are all the same in the column description dataset: "Normalized information about building where the client lives, What is average (_AVG suffix), modus (_MODE suffix), median (_MEDI suffix) apartment size, common area, living area, age of building, number of elevators, number of entrances, state of the building, number of floor",normalized".
    - Well, here it gets trick. I intend to split the data between train and validation. Considering the model will be deployed, will data always come normalized?
    - Some columns, such as ```WALLSMATERIAL_MODE```, are categorical. They look like normal categorical columns and not from the mode (suffix).
- ```DAYS_LAST_PHONE_CHANGE```: called for my attention. Applicants changed their phone in the exact date of the application? Seems odd.
- From column 96 ```FLAG_DOCUMENT_2``` to column 115 ```FLAG_DOCUMENT_21```, there is also room for studying correlation and feature engineering, such as rearrangement.

### bureau.csv

In [40]:
overview_data(
    "bureau.csv", display_option="condensed", sample_mode="random", sample_size=5
)

#### Dataset name (.csv)

bureau.csv


#### Dataset Random Samples

SK_ID_CURR  SK_ID_BUREAU CREDIT_ACTIVE CREDIT_CURRENCY  DAYS_CREDIT  \
1236926      305481       5916012        Closed      currency 1        -1531   
904815       288975       5667606        Closed      currency 1        -2263   
849614       309911       6052617        Closed      currency 1         -932   
16038        358247       6025723        Closed      currency 1         -800   
124732       452117       5242575        Closed      currency 1         -782   

         CREDIT_DAY_OVERDUE  DAYS_CREDIT_ENDDATE  DAYS_ENDDATE_FACT  \
1236926                   0               -435.0             -846.0   
904815                    0              -1166.0            -1165.0   
849614                    0                911.0             -233.0   
16038                     0                -70.0             -636.0   
124732                    0               -417.0             -415.0   

         AMT_CREDIT_MAX_OVERDUE  CNT_CREDIT_PROLONG  AMT_CREDIT_SUM  \
1236926                     NaN                   0        450000.0   
904815                      NaN                   0        169884.0   
849614                      NaN                   0        112500.0   
16038                       NaN                   0        140611.5   
124732                      0.0                   0         43960.5   

         AMT_CREDIT_SUM_DEBT  AMT_CREDIT_SUM_LIMIT  AMT_CREDIT_SUM_OVERDUE  \
1236926                  0.0                   NaN                     0.0   
904815                   0.0                   0.0                     0.0   
849614                   0.0                   NaN                     0.0   
16038                    NaN                   NaN                     0.0   
124732                   0.0                   0.0                     0.0   

             CREDIT_TYPE  DAYS_CREDIT_UPDATE  AMT_ANNUITY  
1236926      Credit card                -846          0.0  
904815   Consumer credit               -1164          NaN  
849614       Credit card                -228          NaN  
16038    Consumer credit                -636          0.0  
124732   Consumer credit                -415          NaN

#### Dataset info

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1716428 entries, 0 to 1716427
Data columns (total 17 columns):
 #   Column                  Dtype  
---  ------                  -----  
 0   SK_ID_CURR              int64  
 1   SK_ID_BUREAU            int64  
 2   CREDIT_ACTIVE           object 
 3   CREDIT_CURRENCY         object 
 4   DAYS_CREDIT             int64  
 5   CREDIT_DAY_OVERDUE      int64  
 6   DAYS_CREDIT_ENDDATE     float64
 7   DAYS_ENDDATE_FACT       float64
 8   AMT_CREDIT_MAX_OVERDUE  float64
 9   CNT_CREDIT_PROLONG      int64  
 10  AMT_CREDIT_SUM          float64
 11  AMT_CREDIT_SUM_DEBT     float64
 12  AMT_CREDIT_SUM_LIMIT    float64
 13  AMT_CREDIT_SUM_OVERDUE  float64
 14  CREDIT_TYPE             object 
 15  DAYS_CREDIT_UPDATE      int64  
 16  AMT_ANNUITY             float64
dtypes: float64(8), int64(6), object(3)
memory usage: 222.6+ MB


#### Dataset descriptive statistics

SK_ID_CURR  SK_ID_BUREAU CREDIT_ACTIVE CREDIT_CURRENCY  \
count   1.716428e+06  1.716428e+06       1716428         1716428   
unique           NaN           NaN             4               4   
top              NaN           NaN        Closed      currency 1   
freq             NaN           NaN       1079273         1715020   
mean    2.782149e+05  5.924434e+06           NaN             NaN   
std     1.029386e+05  5.322657e+05           NaN             NaN   
min     1.000010e+05  5.000000e+06           NaN             NaN   
25%     1.888668e+05  5.463954e+06           NaN             NaN   
50%     2.780550e+05  5.926304e+06           NaN             NaN   
75%     3.674260e+05  6.385681e+06           NaN             NaN   
max     4.562550e+05  6.843457e+06           NaN             NaN   

         DAYS_CREDIT  CREDIT_DAY_OVERDUE  DAYS_CREDIT_ENDDATE  \
count   1.716428e+06        1.716428e+06         1.610875e+06   
unique           NaN                 NaN                  NaN   
top              NaN                 NaN                  NaN   
freq             NaN                 NaN                  NaN   
mean   -1.142108e+03        8.181666e-01         5.105174e+02   
std     7.951649e+02        3.654443e+01         4.994220e+03   
min    -2.922000e+03        0.000000e+00        -4.206000e+04   
25%    -1.666000e+03        0.000000e+00        -1.138000e+03   
50%    -9.870000e+02        0.000000e+00        -3.300000e+02   
75%    -4.740000e+02        0.000000e+00         4.740000e+02   
max     0.000000e+00        2.792000e+03         3.119900e+04   

        DAYS_ENDDATE_FACT  AMT_CREDIT_MAX_OVERDUE  CNT_CREDIT_PROLONG  \
count        1.082775e+06            5.919400e+05        1.716428e+06   
unique                NaN                     NaN                 NaN   
top                   NaN                     NaN                 NaN   
freq                  NaN                     NaN                 NaN   
mean        -1.017437e+03            3.825418e+03        6.410406e-03   
std          7.140106e+02            2.060316e+05        9.622391e-02   
min         -4.202300e+04            0.000000e+00        0.000000e+00   
25%         -1.489000e+03            0.000000e+00        0.000000e+00   
50%         -8.970000e+02            0.000000e+00        0.000000e+00   
75%         -4.250000e+02            0.000000e+00        0.000000e+00   
max          0.000000e+00            1.159872e+08        9.000000e+00   

        AMT_CREDIT_SUM  AMT_CREDIT_SUM_DEBT  AMT_CREDIT_SUM_LIMIT  \
count     1.716415e+06         1.458759e+06          1.124648e+06   
unique             NaN                  NaN                   NaN   
top                NaN                  NaN                   NaN   
freq               NaN                  NaN                   NaN   
mean      3.549946e+05         1.370851e+05          6.229515e+03   
std       1.149811e+06         6.774011e+05          4.503203e+04   
min       0.000000e+00        -4.705600e+06         -5.864061e+05   
25%       5.130000e+04         0.000000e+00          0.000000e+00   
50%       1.255185e+05         0.000000e+00          0.000000e+00   
75%       3.150000e+05         4.015350e+04          0.000000e+00   
max       5.850000e+08         1.701000e+08          4.705600e+06   

        AMT_CREDIT_SUM_OVERDUE      CREDIT_TYPE  DAYS_CREDIT_UPDATE  \
count             1.716428e+06          1716428        1.716428e+06   
unique                     NaN               15                 NaN   
top                        NaN  Consumer credit                 NaN   
freq                       NaN          1251615                 NaN   
mean              3.791276e+01              NaN       -5.937483e+02   
std               5.937650e+03              NaN        7.207473e+02   
min               0.000000e+00              NaN       -4.194700e+04   
25%               0.000000e+00              NaN       -9.080000e+02   
50%               0.000000e+00              NaN       -3.950000e+02

In [47]:
describe_features("bureau.csv", display_option="condensed")

Dataset                  Column  \
122  bureau.csv              SK_ID_CURR   
123  bureau.csv            SK_BUREAU_ID   
124  bureau.csv           CREDIT_ACTIVE   
125  bureau.csv         CREDIT_CURRENCY   
126  bureau.csv             DAYS_CREDIT   
127  bureau.csv      CREDIT_DAY_OVERDUE   
128  bureau.csv     DAYS_CREDIT_ENDDATE   
129  bureau.csv       DAYS_ENDDATE_FACT   
130  bureau.csv  AMT_CREDIT_MAX_OVERDUE   
131  bureau.csv      CNT_CREDIT_PROLONG   
132  bureau.csv          AMT_CREDIT_SUM   
133  bureau.csv     AMT_CREDIT_SUM_DEBT   
134  bureau.csv    AMT_CREDIT_SUM_LIMIT   
135  bureau.csv  AMT_CREDIT_SUM_OVERDUE   
136  bureau.csv             CREDIT_TYPE   
137  bureau.csv      DAYS_CREDIT_UPDATE   
138  bureau.csv             AMT_ANNUITY   

                                           Description  \
122  ID of loan in our sample - one loan in our sam...   
123  Recoded ID of previous Credit Bureau credit re...   
124  Status of the Credit Bureau (CB) reported credits   
125       Recoded currency of the Credit Bureau credit   
126  How many days before current application did c...   
127  Number of days past due on CB credit at the ti...   
128  Remaining duration of CB credit (in days) at t...   
129  Days since CB credit ended at the time of appl...   
130  Maximal amount overdue on the Credit Bureau cr...   
131  How many times was the Credit Bureau credit pr...   
132  Current credit amount for the Credit Bureau cr...   
133               Current debt on Credit Bureau credit   
134  Current credit limit of credit card reported i...   
135     Current amount overdue on Credit Bureau credit   
136       Type of Credit Bureau credit (Car, cash,...)   
137  How many days before loan application did last...   
138                Annuity of the Credit Bureau credit   

                                   Special  
122                                 hashed  
123                                 hashed  
124                                    NaN  
125                                recoded  
126  time only relative to the application  
127                                    NaN  
128  time only relative to the application  
129  time only relative to the application  
130                                    NaN  
131                                    NaN  
132                                    NaN  
133                                    NaN  
134                                    NaN  
135                                    NaN  
136                                    NaN  
137  time only relative to the application  
138                                    NaN

#### Highlights & Ideas

There are **1.1M+** records and **17** columns. A lengthy dataset. The are two main challenges here:

1. This is not out flagship dataset. How can we extract information from this dataset to enrich the data we have from ```application_train.csv```?
2. There are multiple entries for some IDs (every previous loan in the Credit Bureau is stored here). How can we summarize the data to integrate it with the other dataset? Use the most recent, eldest, average, use the mode, or even ignore some columns are all possible ideas.

### bureau_balance.csv

In [45]:
overview_data(
    "bureau_balance.csv",
    display_option="condensed",
    sample_mode="random",
    sample_size=5,
)

#### Dataset name (.csv)

bureau_balance.csv


#### Dataset Random Samples

SK_ID_BUREAU  MONTHS_BALANCE STATUS
10077170       6423524             -10      X
15189524       6059583             -20      0
6921920        6298033             -70      C
2380938        6562465              -2      0
21676529       5929260             -31      C

#### Dataset info

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 27299925 entries, 0 to 27299924
Data columns (total 3 columns):
 #   Column          Dtype 
---  ------          ----- 
 0   SK_ID_BUREAU    int64 
 1   MONTHS_BALANCE  int64 
 2   STATUS          object
dtypes: int64(2), object(1)
memory usage: 624.8+ MB


#### Dataset descriptive statistics

SK_ID_BUREAU  MONTHS_BALANCE    STATUS
count   2.729992e+07    2.729992e+07  27299925
unique           NaN             NaN         8
top              NaN             NaN         C
freq             NaN             NaN  13646993
mean    6.036297e+06   -3.074169e+01       NaN
std     4.923489e+05    2.386451e+01       NaN
min     5.001709e+06   -9.600000e+01       NaN
25%     5.730933e+06   -4.600000e+01       NaN
50%     6.070821e+06   -2.500000e+01       NaN
75%     6.431951e+06   -1.100000e+01       NaN
max     6.842888e+06    0.000000e+00       NaN

In [46]:
describe_features("bureau_balance.csv", display_option="condensed")

Dataset          Column  \
139  bureau_balance.csv    SK_BUREAU_ID   
140  bureau_balance.csv  MONTHS_BALANCE   
141  bureau_balance.csv          STATUS   

                                           Description  \
139  Recoded ID of Credit Bureau credit (unique cod...   
140  Month of balance relative to application date ...   
141  Status of Credit Bureau loan during the month ...   

                                   Special  
139                                 hashed  
140  time only relative to the application  
141                                    NaN

#### Highlights & Ideas

As expected from a balance dataset, there are far too many records: **27.2M**. 

Not all ```SK_ID_BUREAU``` from **bureau.csv** will have balance data and not all balance data is complete (some have only for a few months and some only present value X - for unknown - in their ```STATUS```). So, it will be complicated to extract meaningful data from this dataset and to balance the effort-value tradeoff.

### previous_application.csv

In [48]:
overview_data(
    "previous_application.csv",
    display_option="condensed",
    sample_mode="random",
    sample_size=5,
)

#### Dataset name (.csv)

previous_application.csv


#### Dataset Random Samples

SK_ID_PREV  SK_ID_CURR NAME_CONTRACT_TYPE  AMT_ANNUITY  \
876718      1327604      127360     Consumer loans     5318.100   
1463966     1830718      218834         Cash loans    19737.945   
354631      1258084      104079     Consumer loans     3304.710   
408441      2257201      389250     Consumer loans    10912.545   
37031       1038283      385546         Cash loans          NaN   

         AMT_APPLICATION  AMT_CREDIT  AMT_DOWN_PAYMENT  AMT_GOODS_PRICE  \
876718           27616.5     26082.0            2763.0          27616.5   
1463966         135000.0    165226.5               NaN         135000.0   
354631           36184.5     32562.0            3622.5          36184.5   
408441          240948.0    240948.0               0.0         240948.0   
37031                0.0         0.0               NaN              NaN   

        WEEKDAY_APPR_PROCESS_START  HOUR_APPR_PROCESS_START  ...  \
876718                   WEDNESDAY                       10  ...   
1463966                  WEDNESDAY                        8  ...   
354631                      FRIDAY                       11  ...   
408441                    SATURDAY                        9  ...   
37031                     THURSDAY                       13  ...   

         NAME_SELLER_INDUSTRY  CNT_PAYMENT  NAME_YIELD_GROUP  \
876718   Consumer electronics          6.0              high   
1463966                   XNA         12.0              high   
354631           Connectivity         12.0            middle   
408441   Consumer electronics         24.0        low_action   
37031                     XNA          NaN               XNA   

                    PRODUCT_COMBINATION  DAYS_FIRST_DRAWING DAYS_FIRST_DUE  \
876718      POS household with interest            365243.0        -1174.0   
1463966               Cash Street: high            365243.0        -1193.0   
354631         POS mobile with interest            365243.0         -574.0   
408441   POS household without interest            365243.0         -724.0   
37031                              Cash                 NaN            NaN   

        DAYS_LAST_DUE_1ST_VERSION  DAYS_LAST_DUE DAYS_TERMINATION  \
876718                    -1024.0        -1024.0          -1017.0   
1463966                    -863.0         -863.0           -855.0   
354631                     -244.0         -574.0           -569.0   
408441                      -34.0          -34.0            -26.0   
37031                         NaN            NaN              NaN   

        NFLAG_INSURED_ON_APPROVAL  
876718                        0.0  
1463966                       1.0  
354631                        0.0  
408441                        0.0  
37031                         NaN  

[5 rows x 37 columns]

#### Dataset info

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1670214 entries, 0 to 1670213
Data columns (total 37 columns):
 #   Column                       Non-Null Count    Dtype  
---  ------                       --------------    -----  
 0   SK_ID_PREV                   1670214 non-null  int64  
 1   SK_ID_CURR                   1670214 non-null  int64  
 2   NAME_CONTRACT_TYPE           1670214 non-null  object 
 3   AMT_ANNUITY                  1297979 non-null  float64
 4   AMT_APPLICATION              1670214 non-null  float64
 5   AMT_CREDIT                   1670213 non-null  float64
 6   AMT_DOWN_PAYMENT             774370 non-null   float64
 7   AMT_GOODS_PRICE              1284699 non-null  float64
 8   WEEKDAY_APPR_PROCESS_START   1670214 non-null  object 
 9   HOUR_APPR_PROCESS_START      1670214 non-null  int64  
 10  FLAG_LAST_APPL_PER_CONTRACT  1670214 non-null  object 
 11  NFLAG_LAST_APPL_IN_DAY       1670214 non-null  int64  
 12  RATE_DOWN_PAYMENT            774370 non-nu

#### Dataset descriptive statistics

SK_ID_PREV    SK_ID_CURR NAME_CONTRACT_TYPE   AMT_ANNUITY  \
count   1.670214e+06  1.670214e+06            1670214  1.297979e+06   
unique           NaN           NaN                  4           NaN   
top              NaN           NaN         Cash loans           NaN   
freq             NaN           NaN             747553           NaN   
mean    1.923089e+06  2.783572e+05                NaN  1.595512e+04   
std     5.325980e+05  1.028148e+05                NaN  1.478214e+04   
min     1.000001e+06  1.000010e+05                NaN  0.000000e+00   
25%     1.461857e+06  1.893290e+05                NaN  6.321780e+03   
50%     1.923110e+06  2.787145e+05                NaN  1.125000e+04   
75%     2.384280e+06  3.675140e+05                NaN  2.065842e+04   
max     2.845382e+06  4.562550e+05                NaN  4.180581e+05   

        AMT_APPLICATION    AMT_CREDIT  AMT_DOWN_PAYMENT  AMT_GOODS_PRICE  \
count      1.670214e+06  1.670213e+06      7.743700e+05     1.284699e+06   
unique              NaN           NaN               NaN              NaN   
top                 NaN           NaN               NaN              NaN   
freq                NaN           NaN               NaN              NaN   
mean       1.752339e+05  1.961140e+05      6.697402e+03     2.278473e+05   
std        2.927798e+05  3.185746e+05      2.092150e+04     3.153966e+05   
min        0.000000e+00  0.000000e+00     -9.000000e-01     0.000000e+00   
25%        1.872000e+04  2.416050e+04      0.000000e+00     5.084100e+04   
50%        7.104600e+04  8.054100e+04      1.638000e+03     1.123200e+05   
75%        1.803600e+05  2.164185e+05      7.740000e+03     2.340000e+05   
max        6.905160e+06  6.905160e+06      3.060045e+06     6.905160e+06   

       WEEKDAY_APPR_PROCESS_START  HOUR_APPR_PROCESS_START  ...  \
count                     1670214             1.670214e+06  ...   
unique                          7                      NaN  ...   
top                       TUESDAY                      NaN  ...   
freq                       255118                      NaN  ...   
mean                          NaN             1.248418e+01  ...   
std                           NaN             3.334028e+00  ...   
min                           NaN             0.000000e+00  ...   
25%                           NaN             1.000000e+01  ...   
50%                           NaN             1.200000e+01  ...   
75%                           NaN             1.500000e+01  ...   
max                           NaN             2.300000e+01  ...   

       NAME_SELLER_INDUSTRY   CNT_PAYMENT  NAME_YIELD_GROUP  \
count               1670214  1.297984e+06           1670214   
unique                   11           NaN                 5   
top                     XNA           NaN               XNA   
freq                 855720           NaN            517215   
mean                    NaN  1.605408e+01               NaN   
std                     NaN  1.456729e+01               NaN   
min                     NaN  0.000000e+00               NaN   
25%                     NaN  6.000000e+00               NaN   
50%                     NaN  1.200000e+01               NaN   
75%                     NaN  2.400000e+01               NaN   
max                     NaN  8.400000e+01               NaN   

        PRODUCT_COMBINATION  DAYS_FIRST_DRAWING DAYS_FIRST_DUE  \
count               1669868       997149.000000  997149.000000   
unique                   17                 NaN            NaN   
top                    Cash                 NaN            NaN   
freq                 285990                 NaN            NaN   
mean                    NaN       342209.855039   13826.269337   
std                     NaN        88916.115834   72444.869708   
min                     NaN        -2922.000000   -2892.000000   
25%                     NaN       365243.000000   -1628.000000   
50%                     NaN       365243.000000    -831.000000   
75%                   

In [52]:
describe_features("previous_application.csv", display_option="condensed")

Dataset                       Column  \
173  previous_application.csv                  SK_ID_PREV    
174  previous_application.csv                   SK_ID_CURR   
175  previous_application.csv           NAME_CONTRACT_TYPE   
176  previous_application.csv                  AMT_ANNUITY   
177  previous_application.csv              AMT_APPLICATION   
178  previous_application.csv                   AMT_CREDIT   
179  previous_application.csv             AMT_DOWN_PAYMENT   
180  previous_application.csv              AMT_GOODS_PRICE   
181  previous_application.csv   WEEKDAY_APPR_PROCESS_START   
182  previous_application.csv      HOUR_APPR_PROCESS_START   
183  previous_application.csv  FLAG_LAST_APPL_PER_CONTRACT   
184  previous_application.csv       NFLAG_LAST_APPL_IN_DAY   
185  previous_application.csv             NFLAG_MICRO_CASH   
186  previous_application.csv            RATE_DOWN_PAYMENT   
187  previous_application.csv        RATE_INTEREST_PRIMARY   
188  previous_application.csv     RATE_INTEREST_PRIVILEGED   
189  previous_application.csv       NAME_CASH_LOAN_PURPOSE   
190  previous_application.csv         NAME_CONTRACT_STATUS   
191  previous_application.csv                DAYS_DECISION   
192  previous_application.csv            NAME_PAYMENT_TYPE   
193  previous_application.csv           CODE_REJECT_REASON   
194  previous_application.csv              NAME_TYPE_SUITE   
195  previous_application.csv             NAME_CLIENT_TYPE   
196  previous_application.csv          NAME_GOODS_CATEGORY   
197  previous_application.csv               NAME_PORTFOLIO   
198  previous_application.csv            NAME_PRODUCT_TYPE   
199  previous_application.csv                 CHANNEL_TYPE   
200  previous_application.csv             SELLERPLACE_AREA   
201  previous_application.csv         NAME_SELLER_INDUSTRY   
202  previous_application.csv                  CNT_PAYMENT   
203  previous_application.csv             NAME_YIELD_GROUP   
204  previous_application.csv          PRODUCT_COMBINATION   
205  previous_application.csv           DAYS_FIRST_DRAWING   
206  previous_application.csv               DAYS_FIRST_DUE   
207  previous_application.csv    DAYS_LAST_DUE_1ST_VERSION   
208  previous_application.csv                DAYS_LAST_DUE   
209  previous_application.csv             DAYS_TERMINATION   
210  previous_application.csv    NFLAG_INSURED_ON_APPROVAL   

                                           Description  \
173  ID of previous credit in Home credit related t...   
174                           ID of loan in our sample   
175  Contract product type (Cash loan, consumer loa...   
176                    Annuity of previous application   
177  For how much credit did client ask on the prev...   
178  Final credit amount on the previous applicatio...   
179           Down payment on the previous application   
180  Goods price of good that client asked for (if ...   
181  On which day of the week did the client apply ...   
182  Approximately at what day hour did the client ...   
183  Flag if it was last application for the previo...   
184  Flag if the application was the last applicati...   
185                            Flag Micro finance loan   
186    Down payment rate normalized on previous credit   
187        Interest rate normalized on previous credit   
188        Interest rate normalized on previous credit   
189                           Purpose of the cash loan   
190  Contract status (approved, cancelled, ...) of ...   
191  Relative to current application when was the d...   
192  Payment method that client chose to pay for th...   
193          Why was the previous application rejected   
194  Who accompanied client when applying for the p...   
195  Was the client old or new client when applying...   
196  What kind of goods did the client apply for in...   
197  Was the previous application for CASH, POS, CA...   
198      Was the previous application x-sell o walk-in   
199  Through which channel we acquire

#### Highlights & Ideas

Again, multiple records per ```SK_ID_CURR```. Every previous application the applicant made with Home Credit is stored here. In this case, even more difficult to identify how to extract information from here: we may summarize the data per ```SK_ID_CURR``` using statistical measures or decide which previous application to use.

This Dataset is connected to 3 others, containing installments and more particular data from each ```SK_ID_PREV```.



### POS_CASH_balance.csv

In [51]:
overview_data(
    "POS_CASH_balance.csv",
    display_option="condensed",
    sample_mode="random",
    sample_size=5,
)

#### Dataset name (.csv)

POS_CASH_balance.csv


#### Dataset Random Samples

SK_ID_PREV  SK_ID_CURR  MONTHS_BALANCE  CNT_INSTALMENT  \
6385902     1141316      278001             -37            12.0   
389847      2682740      248899             -67            12.0   
7302029     2136647      209965             -73            18.0   
6010681     1678541      432875              -8            24.0   
2031441     1870173      117636             -86            10.0   

         CNT_INSTALMENT_FUTURE NAME_CONTRACT_STATUS  SK_DPD  SK_DPD_DEF  
6385902                    0.0            Completed       0           0  
389847                     4.0               Active       0           0  
7302029                   18.0               Active       0           0  
6010681                   16.0               Active       0           0  
2031441                    7.0               Active       0           0

#### Dataset info

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10001358 entries, 0 to 10001357
Data columns (total 8 columns):
 #   Column                 Dtype  
---  ------                 -----  
 0   SK_ID_PREV             int64  
 1   SK_ID_CURR             int64  
 2   MONTHS_BALANCE         int64  
 3   CNT_INSTALMENT         float64
 4   CNT_INSTALMENT_FUTURE  float64
 5   NAME_CONTRACT_STATUS   object 
 6   SK_DPD                 int64  
 7   SK_DPD_DEF             int64  
dtypes: float64(2), int64(5), object(1)
memory usage: 610.4+ MB


#### Dataset descriptive statistics

SK_ID_PREV    SK_ID_CURR  MONTHS_BALANCE  CNT_INSTALMENT  \
count   1.000136e+07  1.000136e+07    1.000136e+07    9.975287e+06   
unique           NaN           NaN             NaN             NaN   
top              NaN           NaN             NaN             NaN   
freq             NaN           NaN             NaN             NaN   
mean    1.903217e+06  2.784039e+05   -3.501259e+01    1.708965e+01   
std     5.358465e+05  1.027637e+05    2.606657e+01    1.199506e+01   
min     1.000001e+06  1.000010e+05   -9.600000e+01    1.000000e+00   
25%     1.434405e+06  1.895500e+05   -5.400000e+01    1.000000e+01   
50%     1.896565e+06  2.786540e+05   -2.800000e+01    1.200000e+01   
75%     2.368963e+06  3.674290e+05   -1.300000e+01    2.400000e+01   
max     2.843499e+06  4.562550e+05   -1.000000e+00    9.200000e+01   

        CNT_INSTALMENT_FUTURE NAME_CONTRACT_STATUS        SK_DPD    SK_DPD_DEF  
count            9.975271e+06             10001358  1.000136e+07  1.000136e+07  
unique                    NaN                    9           NaN           NaN  
top                       NaN               Active           NaN           NaN  
freq                      NaN              9151119           NaN           NaN  
mean             1.048384e+01                  NaN  1.160693e+01  6.544684e-01  
std              1.110906e+01                  NaN  1.327140e+02  3.276249e+01  
min              0.000000e+00                  NaN  0.000000e+00  0.000000e+00  
25%              3.000000e+00                  NaN  0.000000e+00  0.000000e+00  
50%              7.000000e+00                  NaN  0.000000e+00  0.000000e+00  
75%              1.400000e+01                  NaN  0.000000e+00  0.000000e+00  
max              8.500000e+01                  NaN  4.231000e+03  3.595000e+03

In [53]:
describe_features("POS_CASH_balance.csv", display_option="condensed")

Dataset                 Column  \
142  POS_CASH_balance.csv            SK_ID_PREV    
143  POS_CASH_balance.csv             SK_ID_CURR   
144  POS_CASH_balance.csv         MONTHS_BALANCE   
145  POS_CASH_balance.csv         CNT_INSTALMENT   
146  POS_CASH_balance.csv  CNT_INSTALMENT_FUTURE   
147  POS_CASH_balance.csv   NAME_CONTRACT_STATUS   
148  POS_CASH_balance.csv                 SK_DPD   
149  POS_CASH_balance.csv             SK_DPD_DEF   

                                           Description  \
142  ID of previous credit in Home Credit related t...   
143                           ID of loan in our sample   
144  Month of balance relative to application date ...   
145     Term of previous credit (can change over time)   
146    Installments left to pay on the previous credit   
147                   Contract status during the month   
148  DPD (days past due) during the month of previo...   
149  DPD during the month with tolerance (debts wit...   

                                   Special  
142                                    NaN  
143                                    NaN  
144  time only relative to the application  
145                                    NaN  
146                                    NaN  
147                                    NaN  
148                                    NaN  
149                                    NaN

#### Highlights & Ideas

Well, seems like every "balance" dataframe will present the same challenge. However, this dataset may provide useful information about any days past due (```SK_DPD```) in the past or present.

### installments_payments.csv

In [88]:
overview_data(
    "installments_payments.csv",
    display_option="condensed",
    sample_mode="random",
    sample_size=5,
)

#### Dataset name (.csv)

installments_payments.csv


#### Dataset Random Samples

SK_ID_PREV  SK_ID_CURR  NUM_INSTALMENT_VERSION  \
5672587     2652367      232137                     0.0   
566432      1845882      167969                     1.0   
1780626     2430317      105881                     1.0   
2601634     1052426      170461                     1.0   
4353865     2148550      265570                     0.0   

         NUM_INSTALMENT_NUMBER  DAYS_INSTALMENT  DAYS_ENTRY_PAYMENT  \
5672587                     18          -2398.0             -2403.0   
566432                      14           -136.0              -140.0   
1780626                     15           -120.0              -120.0   
2601634                      7          -1083.0             -1090.0   
4353865                     47          -1125.0             -1128.0   

         AMT_INSTALMENT  AMT_PAYMENT  
5672587        4500.000     4500.000  
566432         8043.480     8043.480  
1780626       48222.855    48222.855  
2601634       11721.735    11721.735  
4353865        7875.000     7875.000

#### Dataset info

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 13605401 entries, 0 to 13605400
Data columns (total 8 columns):
 #   Column                  Dtype  
---  ------                  -----  
 0   SK_ID_PREV              int64  
 1   SK_ID_CURR              int64  
 2   NUM_INSTALMENT_VERSION  float64
 3   NUM_INSTALMENT_NUMBER   int64  
 4   DAYS_INSTALMENT         float64
 5   DAYS_ENTRY_PAYMENT      float64
 6   AMT_INSTALMENT          float64
 7   AMT_PAYMENT             float64
dtypes: float64(5), int64(3)
memory usage: 830.4 MB


#### Dataset descriptive statistics

SK_ID_PREV    SK_ID_CURR  NUM_INSTALMENT_VERSION  \
count  1.360540e+07  1.360540e+07            1.360540e+07   
mean   1.903365e+06  2.784449e+05            8.566373e-01   
std    5.362029e+05  1.027183e+05            1.035216e+00   
min    1.000001e+06  1.000010e+05            0.000000e+00   
25%    1.434191e+06  1.896390e+05            0.000000e+00   
50%    1.896520e+06  2.786850e+05            1.000000e+00   
75%    2.369094e+06  3.675300e+05            1.000000e+00   
max    2.843499e+06  4.562550e+05            1.780000e+02   

       NUM_INSTALMENT_NUMBER  DAYS_INSTALMENT  DAYS_ENTRY_PAYMENT  \
count           1.360540e+07     1.360540e+07        1.360250e+07   
mean            1.887090e+01    -1.042270e+03       -1.051114e+03   
std             2.666407e+01     8.009463e+02        8.005859e+02   
min             1.000000e+00    -2.922000e+03       -4.921000e+03   
25%             4.000000e+00    -1.654000e+03       -1.662000e+03   
50%             8.000000e+00    -8.180000e+02       -8.270000e+02   
75%             1.900000e+01    -3.610000e+02       -3.700000e+02   
max             2.770000e+02    -1.000000e+00       -1.000000e+00   

       AMT_INSTALMENT   AMT_PAYMENT  
count    1.360540e+07  1.360250e+07  
mean     1.705091e+04  1.723822e+04  
std      5.057025e+04  5.473578e+04  
min      0.000000e+00  0.000000e+00  
25%      4.226085e+03  3.398265e+03  
50%      8.884080e+03  8.125515e+03  
75%      1.671021e+04  1.610842e+04  
max      3.771488e+06  3.771488e+06

In [86]:
describe_features("installments_payments.csv", display_option="condensed")

Dataset                  Column  \
211  installments_payments.csv             SK_ID_PREV    
212  installments_payments.csv              SK_ID_CURR   
213  installments_payments.csv  NUM_INSTALMENT_VERSION   
214  installments_payments.csv   NUM_INSTALMENT_NUMBER   
215  installments_payments.csv         DAYS_INSTALMENT   
216  installments_payments.csv      DAYS_ENTRY_PAYMENT   
217  installments_payments.csv          AMT_INSTALMENT   
218  installments_payments.csv             AMT_PAYMENT   

                                           Description  \
211  ID of previous credit in Home credit related t...   
212                           ID of loan in our sample   
213  Version of installment calendar (0 is for cred...   
214            On which installment we observe payment   
215  When the installment of previous credit was su...   
216  When was the installments of previous credit p...   
217  What was the prescribed installment amount of ...   
218  What the client actually paid on previous cred...   

                                   Special  
211                                 hashed  
212                                 hashed  
213                                    NaN  
214                                    NaN  
215  time only relative to the application  
216  time only relative to the application  
217                                    NaN  
218                                    NaN

#### Highlights & Ideas

A dataset with only numerical features, presenting the same challenge as the other datasets.

### credit_card_balance.csv

In [127]:
overview_data(
    "credit_card_balance.csv",
    display_option="condensed",
    sample_mode="random",
    sample_size=5,
)

#### Dataset name (.csv)

credit_card_balance.csv


#### Dataset Random Samples

SK_ID_PREV  SK_ID_CURR  MONTHS_BALANCE  AMT_BALANCE  \
2515502     2229653      284348             -45   136399.635   
1307940     1786293      156412             -77        0.000   
318549      2199394      397441             -88        0.000   
2558004     1171732      364620              -6   345647.430   
3802735     1798989      361308             -69    93392.505   

         AMT_CREDIT_LIMIT_ACTUAL  AMT_DRAWINGS_ATM_CURRENT  \
2515502                   135000                       0.0   
1307940                    90000                       0.0   
318549                    180000                       0.0   
2558004                   337500                   84600.0   
3802735                   202500                       0.0   

         AMT_DRAWINGS_CURRENT  AMT_DRAWINGS_OTHER_CURRENT  \
2515502                   0.0                         0.0   
1307940                   0.0                         0.0   
318549                    0.0                         0.0   
2558004               84600.0                         0.0   
3802735                   0.0                         0.0   

         AMT_DRAWINGS_POS_CURRENT  AMT_INST_MIN_REGULARITY  ...  \
2515502                       0.0                 6750.000  ...   
1307940                       0.0                    0.000  ...   
318549                        0.0                 9000.000  ...   
2558004                       0.0                16579.215  ...   
3802735                       0.0                10125.000  ...   

         AMT_RECIVABLE  AMT_TOTAL_RECEIVABLE  CNT_DRAWINGS_ATM_CURRENT  \
2515502     136399.635            136399.635                       0.0   
1307940          0.000                 0.000                       0.0   
318549           0.000                 0.000                       0.0   
2558004     335103.930            335103.930                       6.0   
3802735      93392.505             93392.505                       0.0   

         CNT_DRAWINGS_CURRENT  CNT_DRAWINGS_OTHER_CURRENT  \
2515502                     0                         0.0   
1307940                     0                         0.0   
318549                      0                         0.0   
2558004                     6                         0.0   
3802735                     0                         0.0   

         CNT_DRAWINGS_POS_CURRENT  CNT_INSTALMENT_MATURE_CUM  \
2515502                       0.0                       48.0   
1307940                       0.0                        5.0   
318549                        0.0                        1.0   
2558004                       0.0                        2.0   
3802735                       0.0                       18.0   

         NAME_CONTRACT_STATUS  SK_DPD  SK_DPD_DEF  
2515502                Active       1           1  
1307940                Active       0           0  
318549                 Active       0           0  
2558004                Active       0           0  
3802735                Active       0           0  

[5 rows x 23 columns]

#### Dataset info

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3840312 entries, 0 to 3840311
Data columns (total 23 columns):
 #   Column                      Dtype  
---  ------                      -----  
 0   SK_ID_PREV                  int64  
 1   SK_ID_CURR                  int64  
 2   MONTHS_BALANCE              int64  
 3   AMT_BALANCE                 float64
 4   AMT_CREDIT_LIMIT_ACTUAL     int64  
 5   AMT_DRAWINGS_ATM_CURRENT    float64
 6   AMT_DRAWINGS_CURRENT        float64
 7   AMT_DRAWINGS_OTHER_CURRENT  float64
 8   AMT_DRAWINGS_POS_CURRENT    float64
 9   AMT_INST_MIN_REGULARITY     float64
 10  AMT_PAYMENT_CURRENT         float64
 11  AMT_PAYMENT_TOTAL_CURRENT   float64
 12  AMT_RECEIVABLE_PRINCIPAL    float64
 13  AMT_RECIVABLE               float64
 14  AMT_TOTAL_RECEIVABLE        float64
 15  CNT_DRAWINGS_ATM_CURRENT    float64
 16  CNT_DRAWINGS_CURRENT        int64  
 17  CNT_DRAWINGS_OTHER_CURRENT  float64
 18  CNT_DRAWINGS_POS_CURRENT    float64
 19  CNT_INSTALMENT_MATURE

#### Dataset descriptive statistics

SK_ID_PREV    SK_ID_CURR  MONTHS_BALANCE   AMT_BALANCE  \
count   3.840312e+06  3.840312e+06    3.840312e+06  3.840312e+06   
unique           NaN           NaN             NaN           NaN   
top              NaN           NaN             NaN           NaN   
freq             NaN           NaN             NaN           NaN   
mean    1.904504e+06  2.783242e+05   -3.452192e+01  5.830016e+04   
std     5.364695e+05  1.027045e+05    2.666775e+01  1.063070e+05   
min     1.000018e+06  1.000060e+05   -9.600000e+01 -4.202502e+05   
25%     1.434385e+06  1.895170e+05   -5.500000e+01  0.000000e+00   
50%     1.897122e+06  2.783960e+05   -2.800000e+01  0.000000e+00   
75%     2.369328e+06  3.675800e+05   -1.100000e+01  8.904669e+04   
max     2.843496e+06  4.562500e+05   -1.000000e+00  1.505902e+06   

        AMT_CREDIT_LIMIT_ACTUAL  AMT_DRAWINGS_ATM_CURRENT  \
count              3.840312e+06              3.090496e+06   
unique                      NaN                       NaN   
top                         NaN                       NaN   
freq                        NaN                       NaN   
mean               1.538080e+05              5.961325e+03   
std                1.651457e+05              2.822569e+04   
min                0.000000e+00             -6.827310e+03   
25%                4.500000e+04              0.000000e+00   
50%                1.125000e+05              0.000000e+00   
75%                1.800000e+05              0.000000e+00   
max                1.350000e+06              2.115000e+06   

        AMT_DRAWINGS_CURRENT  AMT_DRAWINGS_OTHER_CURRENT  \
count           3.840312e+06                3.090496e+06   
unique                   NaN                         NaN   
top                      NaN                         NaN   
freq                     NaN                         NaN   
mean            7.433388e+03                2.881696e+02   
std             3.384608e+04                8.201989e+03   
min            -6.211620e+03                0.000000e+00   
25%             0.000000e+00                0.000000e+00   
50%             0.000000e+00                0.000000e+00   
75%             0.000000e+00                0.000000e+00   
max             2.287098e+06                1.529847e+06   

        AMT_DRAWINGS_POS_CURRENT  AMT_INST_MIN_REGULARITY  ...  AMT_RECIVABLE  \
count               3.090496e+06             3.535076e+06  ...   3.840312e+06   
unique                       NaN                      NaN  ...            NaN   
top                          NaN                      NaN  ...            NaN   
freq                         NaN                      NaN  ...            NaN   
mean                2.968805e+03             3.540204e+03  ...   5.808881e+04   
std                 2.079689e+04             5.600154e+03  ...   1.059654e+05   
min                 0.000000e+00             0.000000e+00  ...  -4.202502e+05   
25%                 0.000000e+00             0.000000e+00  ...   0.000000e+00   
50%                 0.000000e+00             0.000000e+00  ...   0.000000e+00   
75%                 0.000000e+00             6.633911e+03  ...   8.889949e+04   
max                 2.239274e+06             2.028820e+05  ...   1.493338e+06   

        AMT_TOTAL_RECEIVABLE  CNT_DRAWINGS_ATM_CURRENT  CNT_DRAWINGS_CURRENT  \
count           3.840312e+06              3.090496e+06          3.840312e+06   
unique                   NaN                       NaN                   NaN   
top                      NaN                       NaN                   NaN   
freq                     NaN                       NaN                   NaN   
mean            5.809829e+04              3.094490e-01          7.031439e-01   
std             1.059718e+05              1.100401e+00          3.190347e+00   
min            -4.202502e+05              0.000000e+00          0.000000e+00   
25%             0.000000e+00              0.000000e+00          0.000000e+00   
50%             0.000000e+00              0.00

In [128]:
describe_features("credit_card_balance.csv", display_option="condensed")

Dataset                      Column  \
150  credit_card_balance.csv                 SK_ID_PREV    
151  credit_card_balance.csv                  SK_ID_CURR   
152  credit_card_balance.csv              MONTHS_BALANCE   
153  credit_card_balance.csv                 AMT_BALANCE   
154  credit_card_balance.csv     AMT_CREDIT_LIMIT_ACTUAL   
155  credit_card_balance.csv    AMT_DRAWINGS_ATM_CURRENT   
156  credit_card_balance.csv        AMT_DRAWINGS_CURRENT   
157  credit_card_balance.csv  AMT_DRAWINGS_OTHER_CURRENT   
158  credit_card_balance.csv    AMT_DRAWINGS_POS_CURRENT   
159  credit_card_balance.csv     AMT_INST_MIN_REGULARITY   
160  credit_card_balance.csv         AMT_PAYMENT_CURRENT   
161  credit_card_balance.csv   AMT_PAYMENT_TOTAL_CURRENT   
162  credit_card_balance.csv    AMT_RECEIVABLE_PRINCIPAL   
163  credit_card_balance.csv               AMT_RECIVABLE   
164  credit_card_balance.csv        AMT_TOTAL_RECEIVABLE   
165  credit_card_balance.csv    CNT_DRAWINGS_ATM_CURRENT   
166  credit_card_balance.csv        CNT_DRAWINGS_CURRENT   
167  credit_card_balance.csv  CNT_DRAWINGS_OTHER_CURRENT   
168  credit_card_balance.csv    CNT_DRAWINGS_POS_CURRENT   
169  credit_card_balance.csv   CNT_INSTALMENT_MATURE_CUM   
170  credit_card_balance.csv        NAME_CONTRACT_STATUS   
171  credit_card_balance.csv                      SK_DPD   
172  credit_card_balance.csv                  SK_DPD_DEF   

                                           Description  \
150  ID of previous credit in Home credit related t...   
151                           ID of loan in our sample   
152  Month of balance relative to application date ...   
153        Balance during the month of previous credit   
154  Credit card limit during the month of the prev...   
155  Amount drawing at ATM during the month of the ...   
156  Amount drawing during the month of the previou...   
157  Amount of other drawings during the month of t...   
158  Amount drawing or buying goods during the mont...   
159  Minimal installment for this month of the prev...   
160  How much did the client pay during the month o...   
161  How much did the client pay during the month i...   
162  Amount receivable for principal on the previou...   
163           Amount receivable on the previous credit   
164     Total amount receivable on the previous credit   
165  Number of drawings at ATM during this month on...   
166  Number of drawings during this month on the pr...   
167  Number of other drawings during this month on ...   
168  Number of drawings for goods during this month...   
169  Number of paid installments on the previous cr...   
170  Contract status (active signed,...) on the pre...   
171  DPD (Days past due) during the month on the pr...   
172  DPD (Days past due) during the month with tole...   

                                   Special  
150                                 hashed  
151                                 hashed  
152  time only relative to the application  
153                                    NaN  
154                                    NaN  
155                                    NaN  
156                                    NaN  
157                                    NaN  
158                                    NaN  
159                                    NaN  
160                                    NaN  
161                                    NaN  
162                                    NaN  
163                                    NaN  
164                                    NaN  
165                                    NaN  
166                                    NaN  
167                                    NaN  
168                                    NaN  
169                                    NaN  
170                                    NaN  
171                                    NaN  
172                                    NaN

#### Highlights & Ideas

Here, we pretty much observe the same behavior from **POS_CASH_balance.csv**, but with credit card characteristics.

### Final Thoughts

Through this first analysis, seems like we might have a considerable amount of the Iunfamous **NaNs** in the data. We also have multiple huge datasets with relations that are not so easy to fit as feature data. Therefore, I will dive a little bit deeper into the exploratory data analysis to gather more insights.

*Next notebook: [3.0-ejk-eda-applications](3.0-ejk-eda-applications.ipynb).*